browser/brave_rewards/test/creator_detection_browsertest.cc

```diff
@@ -0,0 +1,349 @@
+    embedded_https_test_server().SetCertHostnames(
+        {"twitter.com", "github.com", "api.github.com", "reddit.com",
+         "www.twitch.tv", "vimeo.com", "www.youtube.com"});
+    embedded_https_test_server().RegisterRequestHandler(base::BindRepeating(
+        &CreatorDetectionBrowserTest::HandleRequest, base::Unretained(this)));
+    ASSERT_TRUE(embedded_https_test_server().Start());
+  }
+
+  void TearDownOnMainThread() override {
+    EXPECT_TRUE(embedded_https_test_server().ShutdownAndWaitUntilComplete());
+    InProcessBrowserTest::TearDownOnMainThread();
+  }
+
+  using RequestCallback =
+      base::RepeatingCallback<void(std::string_view, BasicHttpResponse&)>;
+
+  template <typename F>
+  void SetRequestCallback(F callback) {
+    request_callback_ = base::BindLambdaForTesting(callback);
+  }
+
+  void SetResponseHTML(std::string_view html) {
+    std::string content(html);
+    SetRequestCallback(
+        [content](std::string_view path, BasicHttpResponse& response) {
+          response.set_content(content);
+        });
+  }
+
+  void SetRewardsEnabled() {
+    auto* prefs = browser()->profile()->GetPrefs();
+    prefs->SetBoolean(brave_rewards::prefs::kEnabled, true);
+  }
+
+  void NavigateTo(std::string_view host, std::string_view path) {
+    auto url = embedded_https_test_server().GetURL(host, path);
+    ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), url));
+  }
+
+  RewardsTabHelper& GetRewardsTabHelper() {
+    auto* web_contents = chrome_test_utils::GetActiveWebContents(this);
+    CHECK(web_contents);
+    auto* tab_helper = RewardsTabHelper::FromWebContents(web_contents);
+    CHECK(tab_helper);
+    return *tab_helper;
+  }
+
+  class TabHelperObserver : public RewardsTabHelper::Observer {
+   public:
+    TabHelperObserver(RewardsTabHelper& tab_helper,
+                      base::OnceCallback<void(std::string)> callback)
+        : callback_(std::move(callback)) {
+      tab_helper_observation_.Observe(&tab_helper);
+    }
+
+    ~TabHelperObserver() override = default;
+
+    void OnPublisherForTabUpdated(const std::string& publisher_id) override {
+      if (callback_) {
+        std::move(callback_).Run(publisher_id);
+      }
+    }
+
+   private:
+    base::OnceCallback<void(std::string)> callback_;
+    RewardsTabHelper::Observation tab_helper_observation_{this};
+  };
+
+ private:
+  std::unique_ptr<HttpResponse> HandleRequest(const HttpRequest& request) {
+    auto response = std::make_unique<BasicHttpResponse>();
+    response->set_code(net::HTTP_OK);
+    response->set_content_type("text/html;charset=utf-8");
+    if (request_callback_) {
+      request_callback_.Run(request.relative_url, *response);
+    }
+    return response;
+  }
+
+  base::test::ScopedFeatureList scoped_feature_list_;
+  RequestCallback request_callback_;
+  std::string response_html_;
+};
+
+IN_PROC_BROWSER_TEST_F(CreatorDetectionBrowserTest, GithubDetection) {
+  SetRewardsEnabled();
```

Comment:
We also need to check that it doesn't do anything when rewards is not enabled. Also if you have something that needs to run at the start of every test it should go in SetUp/SetUpOnMainThread

Reply:
it's pretty easy to miss, I think it would be better to check enabled/disabled for each test so `SetRewardsEnabled(true) .... SetRewardsEnabled(false) ...`? We should also test edge cases like navigating before we get the response from the script.

Reply:
Added testing for:

* Feature disabled
* Same-doc navigation
* Navigation away from a platform page
* Non-platform site creator ID
* Storage to the `publisher_info` table

The AC/activity related behavior of `RewardsTabHelper` is currently tested by `RewardsContributionBrowserTest`.